aptos-move/framework/src/natives/cryptography/groups/arkworks_backend.rs

```diff
@@ -0,0 +1,1896 @@
+    pub g2_affine_serialize_uncompressed: InternalGas,
+    pub g2_affine_to_prepared: InternalGasPerArg,
+    pub g2_affine_to_proj: InternalGas,
+    pub g2_proj_add: InternalGasPerArg,
+    pub g2_proj_addassign: InternalGasPerArg,
+    pub g2_proj_double: InternalGas,
+    pub g2_proj_eq: InternalGas,
+    pub g2_proj_generator: InternalGas,
+    pub g2_proj_infinity: InternalGas,
+    pub g2_proj_mul: InternalGasPerArg,
+    pub g2_proj_mulassign: InternalGasPerArg,
+    pub g2_proj_neg: InternalGas,
+    pub g2_proj_rand: InternalGas,
+    pub g2_proj_sub: InternalGas,
+    pub g2_proj_subassign: InternalGas,
+    pub g2_proj_to_affine: InternalGas,
+    pub g2_proj_to_prepared: InternalGasPerArg,
+    pub pairing_product_base: InternalGas,
+    pub pairing_product_per_pair: InternalGasPerArg,
+}
+
+#[derive(Debug, Clone)]
+pub struct GasParameters {
+    pub bls12_381: Bls12381GasParameters,
+}
+
+#[derive(Tid)]
+pub struct Bls12381Context {
+    fr_store: Vec<Fr>,
+    g1_point_store: Vec<G1Projective>,
+    g2_point_store: Vec<G2Projective>,
+    gt_point_store: Vec<Fq12>,
+}
+
+impl Bls12381Context {
+    pub fn new() -> Self {
+        Self {
+            fr_store: vec![],
+            g1_point_store: vec![],
+            g2_point_store: vec![],
+            gt_point_store: vec![],
+        }
+    }
+
+    pub fn add_scalar(&mut self, scalar: Fr) -> usize {
+        let ret = self.fr_store.len();
+        self.fr_store.push(scalar);
+        ret
+    }
+
+    pub fn get_scalar(&self, handle: usize) -> &Fr {
+        self.fr_store.get(handle).unwrap()
+    }
+
+    pub fn add_g1_point(&mut self, p0: G1Projective) -> usize {
+        let ret = self.g1_point_store.len();
+        self.g1_point_store.push(p0);
+        ret
+    }
+
+    pub fn get_g1_point(&self, handle: usize) -> &G1Projective {
+        self.g1_point_store.get(handle).unwrap()
+    }
+
+    pub fn add_g2_point(&mut self, p0: G2Projective) -> usize {
+        let ret = self.g2_point_store.len();
+        self.g2_point_store.push(p0);
+        ret
+    }
+
+    pub fn get_g2_point(&self, handle: usize) -> &G2Projective {
+        self.g2_point_store.get(handle).unwrap()
+    }
+
+    pub fn add_gt_point(&mut self, point: Fq12) -> usize {
+        let ret = self.gt_point_store.len();
+        self.gt_point_store.push(point);
+        ret
+    }
+
+    pub fn get_gt_point(&self, handle: usize) -> &Fq12 {
+        self.gt_point_store.get(handle).unwrap()
+    }
+}
+
+fn serialize_element_uncompressed_internal(
+    gas_params: &GasParameters,
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert_eq!(1, ty_args.len());
+    let type_tag = context
+        .type_to_type_tag(ty_args.get(0).unwrap())?
+        .to_string();
+    let handle = pop_arg!(args, u64) as usize;
+    match type_tag.as_str() {
+        "0x1::groups::BLS12_381_G1" => {
```

Comment:
Eh... let's declare constants for these strings, please?

Reply:
Fixed.